### PR TITLE
TEP-0056: Pipelines in Pipelines [Problem Statement]

### PR DESCRIPTION
**Pull request by jerop**
Today, users can define and execute `Tasks` and `Custom Tasks`
in `Pipelines`. In this TEP, we propose allowing users to define
and execute `Pipelines` in `Pipelines`, alongside `Tasks` and
`Custom Tasks`.

We previously described the motivation for this TEP in:
- https://github.com/tektoncd/community/pull/374
- https://github.com/tektoncd/community/pull/498

In this change, we build on that problem statement to include the
experiences and learnings using the experimental `Pipelines` in
`Pipelines` `Custom Task` over the past several months.

In summary, the motivation for `Pipelines` in `Pipelines` include
composability, reusability, failure strategies, skipping strategies,
data locality & pod overhead and secure software supply chain.

### CODE REVIEW

File: teps/0056-pipelines-in-pipelines.md [hunk 1]
```diff
@@ -2,186 +2,129 @@
 status: proposed
 title: Pipelines in Pipelines
 creation-date: '2021-03-08'
-last-updated: '2021-08-16'
+last-updated: '2022-05-03'
 authors:
 - '@jerop'
+- '@abayer'
```
Comment (jerop) [inline teps/0056-pipelines-in-pipelines.md:8]: @abayer thank you for the contributions! 😻

**Review by lbernick — APPROVED** [review]

**Comment by dibyom** [comment]
I like the framing of the problems this TEP solves!

/approve

File: teps/0056-pipelines-in-pipelines.md [hunk 2]
```diff
@@ -2,283 +2,389 @@
...
+           v                           
+       add-config 
+           |
+           v 
+        deploy
+```
 
-The canonical place for the latest set of instructions (and the likely source
-of this file) is [here](/teps/NNNN-TEP-template/README.md).
+For further information, read [issue][issue-2134] and [issue][issue-4067].
```
Comment (pritidesai) [inline teps/0056-pipelines-in-pipelines.md:109]: NIT: this translates into:

```
For further information, read issue and issue.
```

A minor formating trouble here.

Also, the biggest pain point in https://github.com/tektoncd/pipeline/issues/4067 is the resources but I agree that the `pipelines in pipelines` indeed simplifies the YAML definition.

File: teps/0056-pipelines-in-pipelines.md [hunk 3]
```diff
@@ -2,283 +2,389 @@
...
--->
+Today, a subset of `Tasks` in a `Pipeline` cannot be distributed as a unit of execution within `Pipelines`. As 
+such, users have to specify those related `Tasks` separately and repeatedly across many `Pipelines`.
 
-<!--
-A table of contents is helpful for quickly jumping to sections of a TEP and for
-highlighting any additional information provided beyond the standard TEP
-template.
+As a `Pipeline` author, I need to define a set of `Tasks` as a complete unit of execution that I can share across
+`Pipelines`. For example, I can define and distribute `linting` set of `Tasks` and `testing` set of `Tasks`:
```
Comment (pritidesai) [inline teps/0056-pipelines-in-pipelines.md:117]: I will argue with this particular use case 🙃 

I think as a community if possible, we should advise our users to transform such tasks into a list of steps and make them part of a single task. But at the same time, it depends on how these tasks are configured and executed. The `tasks` being the only sharable unit of execution and a single point of failure, we have to rely on splitting such process into multiple tasks. One alternative here is to offer guidance such that `linting` is a step with `onError` configured to `continue` such that `report-linter-output` is executed regardless.

File: teps/0056-pipelines-in-pipelines.md [hunk 4]
```diff
@@ -2,283 +2,389 @@
...
+For further information, read [issue][issue-2134] and [issue][issue-4067].
 
-[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
--->
+### Failure Strategies
 
-To improve the reusability, composability and failure strategies in Tekton Pipelines, this TEP addresses defining and executing Pipelines in Pipelines. This TEP scopes the problem, describes the use cases, and identifies the goals and constrains/requirements for the solution.
+Today, when a `Task` in a `Branch` fails, it stops the execution of unrelated `Branches` and the `Pipeline` as a whole.
+When `Tasks` are specified in independent `Branches`, there are no dependencies between them, so users may expect that
+a failure in one `Branch` would not stop the execution of the other `Branch`.
```
Comment (pritidesai) [inline teps/0056-pipelines-in-pipelines.md:153]: This is the biggest and foremost important use case for pipelines in pipelines.

File: teps/0056-pipelines-in-pipelines.md [hunk 5]
```diff
@@ -2,283 +2,389 @@
...
+[doc]: https://docs.google.com/document/d/14Uf7XQEnkMFBpNYRZiwo4dwRfW6do--m3yPhXHx4ybk/edit
+[experiment-proposal]: https://github.com/tektoncd/community/issues/330
+[issue-4067]: https://github.com/tektoncd/pipeline/issues/4067
+[issue-2134]: https://github.com/tektoncd/pipeline/issues/2134
+[when]: https://github.com/tektoncd/pipeline/blob/main/docs/pipelines.md#guarding-a-task-and-its-dependent-tasks
+[slack]: ../teps/images/0056-slack-thread.png
+[slsa]: https://slsa.dev/
+[scorecard]: https://github.com/ossf/scorecard
+[codeql]: https://github.com/github/codeql
+[matrix-uc]: https://github.com/tektoncd/community/pull/600#pullrequestreview-851817251
```
Comment (pritidesai) [inline teps/0056-pipelines-in-pipelines.md:390]: NIT: formatting issue:

This list is not showing at all 🤔 

Screen Shot 2022-05-06 at 10 20 21 AM

Comment (pritidesai) [reply teps/0056-pipelines-in-pipelines.md:390]: missing `-`?

**Review by pritidesai — APPROVED** [review]
thank you @jerop and @abayer 🙏 🎉 

I will request updating the use case in reusability. We have a lot of compelling use cases already but here some of them are overlapping, for example, the composability, reusability, and skipping strategies use case are interchangeable.  

Use case in https://github.com/tektoncd/pipeline/issues/2134#issuecomment-631552148 for reusability can be considered instead.

File: teps/0056-pipelines-in-pipelines.md (hunk 2 above)
Comment (jerop) [reply teps/0056-pipelines-in-pipelines.md]: fixed the formatting, and agreed that it's a resource issue but pipelines in pipelines can help with the composition

File: teps/0056-pipelines-in-pipelines.md (hunk 3 above)
Comment (jerop) [reply teps/0056-pipelines-in-pipelines.md]: makes sense, updated the use cases to the user-provided example in https://github.com/tektoncd/pipeline/issues/2134#issuecomment-631552148, thank you!

File: teps/0056-pipelines-in-pipelines.md (hunk 5 above)
Comment (jerop) [reply teps/0056-pipelines-in-pipelines.md:388]: It's intentional that the lower list doesn't show in markdown - it's a way to provide reusable links in markdown (https://grantwinney.com/cool-markdown-tricks-for-github/#create-reusable-links)

File: teps/0056-pipelines-in-pipelines.md (hunk 3 above)
Comment (pritidesai) [reply teps/0056-pipelines-in-pipelines.md]: thank you @jerop 👍

File: teps/0056-pipelines-in-pipelines.md (hunk 5 above)
Comment (pritidesai) [reply teps/0056-pipelines-in-pipelines.md:388]: great tip, thanks @jerop 👍 

I was looking for the reference to the slack image you have attached and didn't find it in the markdown viewer. Its a great idea to include such screen shots so that the conversation is saved in the repo itself. Without having that reference, its not clearly visible.

This is not a blocker though please merge this!

**Comment by pritidesai** [comment]
thank you @jerop for addressing my comments, please merge!
